pkg/kubectl/deployment.go

```diff
@@ -27,24 +27,41 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
-// DeploymentBasicGeneratorV1 supports stable generation of a deployment
-type DeploymentBasicGeneratorV1 struct {
+// baseDeploymentGenerator: implement the common functionality of
+// DeploymentBasicGeneratorV1 and DeploymentBasicAppsGeneratorV1. To reduce
+// confusion, it's best to keep this struct in the same file as those
+// generators.
+type baseDeploymentGenerator struct {
 	Name   string
 	Images []string
 }
 
-// Ensure it supports the generator pattern that uses parameters specified during construction
-var _ StructuredGenerator = &DeploymentBasicGeneratorV1{}
-
-func (DeploymentBasicGeneratorV1) ParamNames() []GeneratorParam {
+// ParamNames: return the parameters expected by the baseDeploymentGenerator.
+// This method is here to aid in validation. When given a Generator, you can
+// learn what it expects by calling this method.
+func (baseDeploymentGenerator) ParamNames() []GeneratorParam {
 	return []GeneratorParam{
 		{"name", true},
 		{"image", true},
 	}
 }
 
-func (s DeploymentBasicGeneratorV1) Generate(params map[string]interface{}) (runtime.Object, error) {
-	err := ValidateParams(s.ParamNames(), params)
+// validate: check if the caller has forgotten to set one of our fields.
+func (b baseDeploymentGenerator) validate() error {
+	if len(b.Name) == 0 {
+		return fmt.Errorf("name must be specified")
+	}
+	if len(b.Images) == 0 {
+		return fmt.Errorf("at least one image must be specified")
+	}
+	return nil
+}
+
+// fromParams: return a new baseDeploymentGenerator with its parameters parsed
+// from `params`. The current baseDeploymentGenerator (the method receiver) is
```

Comment:
Ah, maybe I should have said the **state** of the current `baseDeploymentGenerator` is unused. You could just as easily instantiate a new `baseDeploymentGenerator` and call ParamNames() on it.